lib/puppet/functions/dig.rb

```diff
@@ -0,0 +1,39 @@
+# Returns a value for a sequence of given keys/indexes into a structure.
+# This function is used to "dig into" a complex data structure by
+# using a sequence of keys / indexes to access a value from which
+# the next key/index is accessed recursively.
+#
+# The first encountered `undef` value or key stops the "dig" and `undef` is returned.
+#
+# An error is raised if an attempt is made to "dig" into
+# something other than an `undef` (which immediately returns `undef`), an Array or a Hash. 
```

Comment:
Support for that is ok to add. If it is an Object, then if the key is an attribute - proceed if not return undef. If attribute exists and is not a Collection, or Object (or undef) error. Would work fine I think. (Needs a separate ticket "Add dig support to Objects").